File: src/Java/myservicebus-azure-service-bus/src/main/java/com/myservicebus/azure/servicebus/AzureServiceBusTransportFactory.java
package com.myservicebus.azure.servicebus;

import com.azure.core.exception.ResourceExistsException;
import com.azure.messaging.servicebus.ServiceBusClientBuilder;
import com.azure.messaging.servicebus.ServiceBusProcessorClient;
import com.azure.messaging.servicebus.ServiceBusSenderClient;
import com.azure.messaging.servicebus.administration.ServiceBusAdministrationClient;
import com.azure.messaging.servicebus.administration.ServiceBusAdministrationClientBuilder;
import com.azure.messaging.servicebus.administration.models.CreateQueueOptions;
import com.azure.messaging.servicebus.administration.models.CreateSubscriptionOptions;
import com.myservicebus.ReceiveTransport;
import com.myservicebus.SendTransport;
import com.myservicebus.TransportCapabilityDescriptor;
import com.myservicebus.TransportCapabilityDescriptors;
import com.myservicebus.TransportFactory;
import com.myservicebus.TransportMessage;
import com.myservicebus.logging.LoggerFactory;
import com.myservicebus.topology.MessageBinding;
import com.myservicebus.topology.ReceiveEndpointTransportTopology;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;

public final class AzureServiceBusTransportFactory implements TransportFactory {
    private final String connectionString;
    private final ServiceBusAdministrationClient administrationClient;
    private final AzureServiceBusTopologyMode topologyMode;
    private final int defaultPrefetchCount;
    private final URI baseAddress;
    private final LoggerFactory loggerFactory;
    private final ConcurrentHashMap<String, SendTransport> sendTransports = new ConcurrentHashMap<>();

    public AzureServiceBusTransportFactory(
            AzureServiceBusFactoryConfigurator configurator,
            LoggerFactory loggerFactory) {
        this.connectionString = configurator.getConnectionString();
        this.topologyMode = configurator.getTopologyMode();
        this.defaultPrefetchCount = configurator.getPrefetchCount();
        this.baseAddress = endpoint(connectionString);
        this.loggerFactory = loggerFactory;
        this.administrationClient = topologyMode == AzureServiceBusTopologyMode.CREATE
                ? new ServiceBusAdministrationClientBuilder()
                        .connectionString(configurator.getManagementConnectionString() != null
                                ? configurator.getManagementConnectionString()
                                : connectionString)
                        .buildClient()
                : null;
    }

    @Override
    public TransportCapabilityDescriptor getCapabilities() {
        return TransportCapabilityDescriptors.AZURE_SERVICE_BUS;
    }

    @Override
    public SendTransport getSendTransport(URI address) {
        AzureServiceBusEndpointAddress destination = AzureServiceBusEndpointAddress.parse(address);
        String key = destination.kind() + ":" + destination.entityName();
        return sendTransports.computeIfAbsent(key, ignored -> new AzureServiceBusSendTransport(
                sender(destination.entityName(), destination.kind()), destination.entityName()));
    }

    @Override
    public ReceiveTransport createReceiveTransport(
            ReceiveEndpointTransportTopology topology,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered) {
        AzureServiceBusReceiveEndpointTopology projected =
                AzureServiceBusReceiveEndpointTopology.project(topology);
        if (topologyMode == AzureServiceBusTopologyMode.CREATE) {
            try {
                ensureTopology(projected);
            } catch (AzureServiceBusTransportException exception) {
                throw exception;
            } catch (Exception exception) {
                throw new AzureServiceBusTransportException(
                        "provision topology", projected.queueName(), exception);
            }
        }

        int prefetchCount = projected.prefetchCount() > 0
                ? projected.prefetchCount()
                : defaultPrefetchCount;
        AtomicReference<AzureServiceBusReceiveTransport> transportReference = new AtomicReference<>();
        ServiceBusProcessorClient processor = new ServiceBusClientBuilder()
                .connectionString(connectionString)
                .processor()
                .queueName(projected.queueName())
                .disableAutoComplete()
                .maxConcurrentCalls(1)
                .prefetchCount(prefetchCount)
                .processMessage(context -> transportReference.get().process(context))
                .processError(context -> transportReference.get().processError(context))
                .buildProcessorClient();
        AzureServiceBusReceiveTransport transport = new AzureServiceBusReceiveTransport(
                processor,
                sender(projected.queueName() + "_skipped", AzureServiceBusEndpointAddress.EntityKind.QUEUE),
                projected.queueName(),
                handler,
                isMessageTypeRegistered,
                projected.temporary() ? null : getFaultAddress(projected.queueName()),
                loggerFactory);
        transportReference.set(transport);
        return transport;
    }

    @Override
    public String getPublishAddress(String entityName) {
        return address(entityName, true, null).toString();
    }

    @Override
    public String getErrorAddress(String endpointName) {
        return address(endpointName + "_error", false, null).toString();
    }

    @Override
    public String getFaultAddress(String endpointName) {
        return address(endpointName + "_fault", true, null).toString();
    }

    @Override
    public String getSendAddress(String queue) {
        return address(queue, false, null).toString();
    }

    public String getTemporaryEndpointAddress(String endpointName) {
        return address(endpointName, false, "temporary=true").toString();
    }

    private ServiceBusSenderClient sender(
            String entityName,
            AzureServiceBusEndpointAddress.EntityKind kind) {
        ServiceBusClientBuilder.ServiceBusSenderClientBuilder builder = new ServiceBusClientBuilder()
                .connectionString(connectionString)
                .sender();
        if (kind == AzureServiceBusEndpointAddress.EntityKind.TOPIC) {
            builder.topicName(entityName);
        } else {
            builder.queueName(entityName);
        }
        return builder.buildClient();
    }

    private void ensureTopology(AzureServiceBusReceiveEndpointTopology topology) {
        ensureQueue(topology.queueName(), topology.temporary());
        if (!topology.temporary()) {
            ensureQueue(topology.queueName() + "_error", false);
            ensureQueue(topology.queueName() + "_skipped", false);
            ensureTopic(topology.queueName() + "_fault");
        }
        for (MessageBinding binding : topology.bindings()) {
            ensureTopic(binding.getEntityName());
            if (!administrationClient.getSubscriptionExists(binding.getEntityName(), topology.queueName())) {
                try {
                    administrationClient.createSubscription(
                            binding.getEntityName(),
                            topology.queueName(),
                            new CreateSubscriptionOptions().setForwardTo(topology.queueName()));
                } catch (ResourceExistsException ignored) {
                    // Another bus instance provisioned the same subscription concurrently.
                }
            }
        }
    }

    private void ensureQueue(String name, boolean temporary) {
        if (administrationClient.getQueueExists(name)) {
            return;
        }
        CreateQueueOptions options = new CreateQueueOptions();
        if (temporary) {
            options.setAutoDeleteOnIdle(Duration.ofMinutes(5));
        }
        try {
            administrationClient.createQueue(name, options);
        } catch (ResourceExistsException ignored) {
            // Another bus instance provisioned the same queue concurrently.
        }
    }

    private void ensureTopic(String name) {
        if (!administrationClient.getTopicExists(name)) {
            try {
                administrationClient.createTopic(name);
            } catch (ResourceExistsException ignored) {
                // Another bus instance provisioned the same topic concurrently.
            }
        }
    }

    private URI address(String entityName, boolean topic, String extraQuery) {
        if (entityName == null || entityName.isBlank()) {
            throw new IllegalArgumentException("Azure Service Bus entity name cannot be blank");
        }
        String query = topic ? "type=topic" : "";
        if (extraQuery != null && !extraQuery.isBlank()) {
            query = query.isEmpty() ? extraQuery : query + "&" + extraQuery;
        }
        String encoded = URLEncoder.encode(entityName, StandardCharsets.UTF_8).replace("+", "%20");
        return URI.create(baseAddress.toString() + encoded + (query.isEmpty() ? "" : "?" + query));
    }

    static URI endpoint(String connectionString) {
        for (String component : connectionString.split(";")) {
            String[] pair = component.split("=", 2);
            if (pair.length == 2 && pair[0].equalsIgnoreCase("Endpoint")) {
                URI endpoint = URI.create(pair[1]);
                return URI.create(endpoint.getScheme() + "://" + endpoint.getAuthority() + "/");
            }
        }
        throw new IllegalArgumentException("Azure Service Bus connection string does not contain Endpoint");
    }
}

File: src/MyServiceBus.AzureServiceBus/AzureServiceBusTransportFactory.cs
using System.Collections.Concurrent;
using Azure.Messaging.ServiceBus;
using Azure.Messaging.ServiceBus.Administration;
using Microsoft.Extensions.Logging;
using MyServiceBus.AzureServiceBus;
using MyServiceBus.Serialization;
using MyServiceBus.Topology;

namespace MyServiceBus;

public sealed class AzureServiceBusTransportFactory : ITransportFactory
{
    private readonly ServiceBusClient _client;
    private readonly ServiceBusAdministrationClient? _administrationClient;
    private readonly AzureServiceBusTopologyMode _topologyMode;
    private readonly int _defaultPrefetchCount;
    private readonly Func<Type, string> _entityNameResolver;
    private readonly Func<string, string> _temporaryEndpointNameFormatter;
    private readonly Uri _baseAddress;
    private readonly ILoggerFactory? _loggerFactory;
    private readonly IInboundMessageResolver _inboundMessageResolver;
    private readonly ConcurrentDictionary<string, ISendTransport> _sendTransports = new(StringComparer.Ordinal);

    public AzureServiceBusTransportFactory(
        ServiceBusClient client,
        IAzureServiceBusFactoryConfigurator configurator,
        IInboundMessageResolver? inboundMessageResolver = null,
        ILoggerFactory? loggerFactory = null)
    {
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(configurator);
        _client = client;
        _topologyMode = configurator.TopologyMode;
        _defaultPrefetchCount = configurator.PrefetchCount;
        _entityNameResolver = configurator.GetEntityName;
        _temporaryEndpointNameFormatter = configurator.TemporaryEndpointNameFormatter;
        _baseAddress = GetEndpoint(configurator.ConnectionString);
        _loggerFactory = loggerFactory;
        _inboundMessageResolver = inboundMessageResolver ?? new InboundMessageResolver();
        if (_topologyMode == AzureServiceBusTopologyMode.Create)
        {
            _administrationClient = new ServiceBusAdministrationClient(
                configurator.ManagementConnectionString ?? configurator.ConnectionString);
        }
    }

    public TransportCapabilityDescriptor Capabilities => TransportCapabilityDescriptors.AzureServiceBus;

    public string GetPublishEntityName(Type messageType) => _entityNameResolver(messageType);

    public Uri GetPublishAddress(string entityName) => CreateAddress(entityName, topic: true);

    public Uri GetPublishAddress(Type messageType) => GetPublishAddress(GetPublishEntityName(messageType));

    public Uri GetTemporaryEndpointAddress(string endpointName) =>
        CreateAddress(FormatTemporaryEndpointName(endpointName), topic: false, "temporary=true");

    public Uri GetErrorAddress(string endpointName) => CreateAddress(endpointName + "_error", topic: false);

    public Uri GetFaultAddress(string endpointName) => CreateAddress(endpointName + "_fault", topic: true);

    public Task<ISendTransport> GetSendTransport(
        Uri address,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(address);
        cancellationToken.ThrowIfCancellationRequested();
        var destination = AzureServiceBusEndpointAddress.Parse(address);
        var key = $"{destination.Kind}:{destination.EntityName}";
        var transport = _sendTransports.GetOrAdd(
            key,
            _ => new AzureServiceBusSendTransport(
                _client.CreateSender(destination.EntityName),
                destination.EntityName));
        return Task.FromResult(transport);
    }

    public async Task<IReceiveTransport> CreateReceiveTransport(
        ReceiveEndpointTransportTopology topology,
        Func<ReceiveContext, Task> handler,
        Func<string?, bool>? isMessageTypeRegistered = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(handler);
        var projected = AzureServiceBusReceiveEndpointTopology.Project(MapTemporaryEndpoint(topology));
        if (_topologyMode == AzureServiceBusTopologyMode.Create)
        {
            try
            {
                await EnsureTopology(projected, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception exception) when (exception is not OperationCanceledException
                                               and not AzureServiceBusTransportException)
            {
                throw new AzureServiceBusTransportException(
                    "provision topology",
                    projected.QueueName,
                    exception);
            }
        }

        var prefetchCount = projected.PrefetchCount > 0 ? projected.PrefetchCount : _defaultPrefetchCount;
        var processor = _client.CreateProcessor(
            projected.QueueName,
            new ServiceBusProcessorOptions
            {
                AutoCompleteMessages = false,
                MaxConcurrentCalls = projected.ConcurrentMessageLimit,
                PrefetchCount = prefetchCount
            });

        return new AzureServiceBusReceiveTransport(
            processor,
            _client.CreateSender(projected.QueueName + "_skipped"),
            projected.QueueName,
            handler,
            isMessageTypeRegistered,
            projected.Temporary ? null : GetErrorAddress(projected.QueueName),
            projected.Temporary ? null : GetFaultAddress(projected.QueueName),
            _inboundMessageResolver,
            _loggerFactory?.CreateLogger<AzureServiceBusReceiveTransport>());
    }

    private async Task EnsureTopology(
        AzureServiceBusReceiveEndpointTopology topology,
        CancellationToken cancellationToken)
    {
        var administrationClient = _administrationClient
            ?? throw new InvalidOperationException("Azure Service Bus administration client is not configured.");

        await EnsureQueue(administrationClient, topology.QueueName, topology.Temporary, cancellationToken)
            .ConfigureAwait(false);

        if (!topology.Temporary)
        {
            await EnsureQueue(administrationClient, topology.QueueName + "_error", false, cancellationToken)
                .ConfigureAwait(false);
            await EnsureQueue(administrationClient, topology.QueueName + "_skipped", false, cancellationToken)
                .ConfigureAwait(false);
            await EnsureTopic(administrationClient, topology.QueueName + "_fault", cancellationToken)
                .ConfigureAwait(false);
        }

        foreach (var binding in topology.Temporary ? [] : topology.Bindings)
        {
            await EnsureTopic(administrationClient, binding.EntityName, cancellationToken).ConfigureAwait(false);
            await EnsureSubscription(
                    administrationClient,
                    binding.EntityName,
                    topology.QueueName,
                    topology.QueueName,
                    cancellationToken)
                .ConfigureAwait(false);
        }
    }

    private ReceiveEndpointTransportTopology MapTemporaryEndpoint(ReceiveEndpointTransportTopology topology)
    {
        if (!topology.Temporary)
            return topology;

        return new ReceiveEndpointTransportTopology(
            FormatTemporaryEndpointName(topology.Name),
            topology.Durable,
            topology.Temporary,
            topology.PrefetchCount,
            topology.Bindings,
            topology.TransportOptions,
            topology.ConcurrentMessageLimit);
    }

    private string FormatTemporaryEndpointName(string endpointName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(endpointName);
        return _temporaryEndpointNameFormatter(endpointName);
    }

    private static async Task EnsureQueue(
        ServiceBusAdministrationClient client,
        string name,
        bool temporary,
        CancellationToken cancellationToken)
    {
        if ((await client.QueueExistsAsync(name, cancellationToken).ConfigureAwait(false)).Value)
            return;

        var options = new CreateQueueOptions(name);
        if (temporary)
            options.AutoDeleteOnIdle = TimeSpan.FromMinutes(5);

        try
        {
            await client.CreateQueueAsync(options, cancellationToken).ConfigureAwait(false);
        }
        catch (ServiceBusException exception) when (exception.Reason == ServiceBusFailureReason.MessagingEntityAlreadyExists)
        {
        }
    }

    private static async Task EnsureTopic(
        ServiceBusAdministrationClient client,
        string name,
        CancellationToken cancellationToken)
    {
        if ((await client.TopicExistsAsync(name, cancellationToken).ConfigureAwait(false)).Value)
            return;

        try
        {
            await client.CreateTopicAsync(name, cancellationToken).ConfigureAwait(false);
        }
        catch (ServiceBusException exception) when (exception.Reason == ServiceBusFailureReason.MessagingEntityAlreadyExists)
        {
        }
    }

    private static async Task EnsureSubscription(
        ServiceBusAdministrationClient client,
        string topic,
        string subscription,
        string forwardTo,
        CancellationToken cancellationToken)
    {
        if ((await client.SubscriptionExistsAsync(topic, subscription, cancellationToken).ConfigureAwait(false)).Value)
            return;

        var options = new CreateSubscriptionOptions(topic, subscription)
        {
            ForwardTo = forwardTo
        };
        try
        {
            await client.CreateSubscriptionAsync(options, cancellationToken).ConfigureAwait(false);
        }
        catch (ServiceBusException exception) when (exception.Reason == ServiceBusFailureReason.MessagingEntityAlreadyExists)
        {
        }
    }

    private Uri CreateAddress(string entityName, bool topic, string? extraQuery = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(entityName);
        var builder = new UriBuilder(new Uri(_baseAddress, Uri.EscapeDataString(entityName)));
        var query = topic ? "type=topic" : string.Empty;
        if (!string.IsNullOrWhiteSpace(extraQuery))
            query = string.IsNullOrEmpty(query) ? extraQuery : query + "&" + extraQuery;
        builder.Query = query;
        return builder.Uri;
    }

    internal static Uri GetEndpoint(string connectionString)
    {
        var properties = ServiceBusConnectionStringProperties.Parse(connectionString);
        var endpoint = properties.Endpoint;
        var builder = new UriBuilder(endpoint)
        {
            Path = "/",
            Query = string.Empty,
            Fragment = string.Empty
        };
        return builder.Uri;
    }
}

internal enum AzureServiceBusEntityKind
{
    Queue,
    Topic
}

internal readonly record struct AzureServiceBusEndpointAddress(
    string EntityName,
    AzureServiceBusEntityKind Kind)
{
    public static AzureServiceBusEndpointAddress Parse(Uri address)
    {
        string entityName;
        AzureServiceBusEntityKind kind;
        if (address.Scheme.Equals("queue", StringComparison.OrdinalIgnoreCase))
        {
            entityName = ParseLogical(address, "queue:");
            kind = AzureServiceBusEntityKind.Queue;
        }
        else if (address.Scheme.Equals("topic", StringComparison.OrdinalIgnoreCase)
                 || address.Scheme.Equals("exchange", StringComparison.OrdinalIgnoreCase))
        {
            entityName = ParseLogical(address, address.Scheme + ":");
            kind = AzureServiceBusEntityKind.Topic;
        }
        else if (address.Scheme.Equals("sb", StringComparison.OrdinalIgnoreCase))
        {
            entityName = Uri.UnescapeDataString(address.AbsolutePath.Trim('/'));
            var type = QueryValue(address, "type");
            kind = type?.ToLowerInvariant() switch
            {
                null or "" => AzureServiceBusEntityKind.Queue,
                "topic" => AzureServiceBusEntityKind.Topic,
                _ => throw new ArgumentException(
                    $"Azure Service Bus entity type '{type}' is not supported.", nameof(address))
            };
        }
        else
        {
            throw new ArgumentException(
                $"Azure Service Bus address scheme '{address.Scheme}' is not supported.", nameof(address));
        }

        if (string.IsNullOrWhiteSpace(entityName))
            throw new ArgumentException("Azure Service Bus entity name cannot be blank.", nameof(address));

        return new AzureServiceBusEndpointAddress(entityName, kind);
    }

    private static string ParseLogical(Uri address, string prefix)
    {
        var value = address.OriginalString[prefix.Length..];
        var queryIndex = value.IndexOf('?');
        return queryIndex >= 0 ? value[..queryIndex] : value;
    }

    private static string? QueryValue(Uri address, string key)
    {
        foreach (var item in address.Query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var pair = item.Split('=', 2);
            if (pair.Length == 2 && pair[0].Equals(key, StringComparison.OrdinalIgnoreCase))
                return Uri.UnescapeDataString(pair[1]);
        }

        return null;
    }
}
